src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1393,29 +1333,249 @@ RewriteResponse TheoryStringsRewriter::preRewrite(TNode node) {
+      {
+        // take up to the end of the string
+        Node ret = NodeManager::currentNM()->mkConst(
+            ::CVC4::String(s.suffix(s.size() - start)));
+        return returnRewrite(node, ret, "ss-const-len-max-oob");
+      }
+      else if (node[2].getConst<Rational>().sgn() <= 0)
+      {
+        Node ret = NodeManager::currentNM()->mkConst(::CVC4::String(""));
+        return returnRewrite(node, ret, "ss-const-len-non-pos");
+      }
+      else
+      {
+        unsigned len =
+            node[2].getConst<Rational>().getNumerator().toUnsignedInt();
+        if (start + len >= s.size())
+        {
+          // take up to the end of the string
+          Node ret = NodeManager::currentNM()->mkConst(
+              ::CVC4::String(s.suffix(s.size() - start)));
+          return returnRewrite(node, ret, "ss-const-end-oob");
+        }
+        else
+        {
+          // compute the substr using the constant string
+          Node ret = NodeManager::currentNM()->mkConst(
+              ::CVC4::String(s.substr(start, len)));
+          return returnRewrite(node, ret, "ss-const-ss");
+        }
+      }
+    }
+  }
+  Node zero = NodeManager::currentNM()->mkConst(CVC4::Rational(0));
+
+  // if entailed non-positive length or negative start point
+  if (checkEntailArith(zero, node[1], true))
+  {
+    Node ret = NodeManager::currentNM()->mkConst(::CVC4::String(""));
+    return returnRewrite(node, ret, "ss-start-neg");
+  }
+  else if (checkEntailArith(zero, node[2]))
+  {
+    Node ret = NodeManager::currentNM()->mkConst(::CVC4::String(""));
+    return returnRewrite(node, ret, "ss-len-non-pos");
+  }
+
+  std::vector<Node> n1;
+  getConcat(node[0], n1);
+
+  // definite inclusion
+  if (node[1] == zero)
+  {
+    Node curr = node[2];
+    std::vector<Node> childrenr;
+    if (stripSymbolicLength(n1, childrenr, 1, curr))
+    {
+      if (curr != zero)
+      {
+        childrenr.push_back(
+            NodeManager::currentNM()->mkNode(kind::STRING_SUBSTR,
+                                             mkConcat(kind::STRING_CONCAT, n1),
+                                             node[1],
+                                             curr));
+      }
+      Node ret = mkConcat(kind::STRING_CONCAT, childrenr);
+      return returnRewrite(node, ret, "ss-len-include");
+    }
+  }
+
+  // symbolic length analysis
+  for (unsigned r = 0; r < 2; r++)
+  {
+    // the amount of characters we can strip
+    Node curr;
+    if (r == 0)
+    {
+      if (node[1] != zero)
+      {
+        // strip up to start point off the start of the string
+        curr = node[1];
+      }
+    }
+    else if (r == 1)
+    {
+      Node tot_len = Rewriter::rewrite(
+          NodeManager::currentNM()->mkNode(kind::STRING_LENGTH, node[0]));
+      Node end_pt = Rewriter::rewrite(
+          NodeManager::currentNM()->mkNode(kind::PLUS, node[1], node[2]));
+      if (node[2] != tot_len)
+      {
+        if (checkEntailArith(node[2], tot_len))
```

Comment:
This could probably be strict to avoid unnecessary work.

Reply:
I think this needs to stay non-strict, since this may make the condition succeed more often.  (Say somehow it is the case that node[2] = tot_len always holds).

Reply:
But if they are the same, then the rewrite has no effect, right?

Reply:
this can happen if the terms are different but entailed equal, for example:

str.len( str.substr( str.++( x, "abcde" ), 1, 5 ) )  and str.len( str.substr( str.++( x, "abcde"), 0, 5 ) )

Reply:
I see, that makes sense.